pkg/dispatcher/dispatcher.go

```diff
@@ -257,13 +260,23 @@ func Stop() {
 	}
 
 	close(getDispatcher().stopChan)
-	maxTimeout := 5
-	for getDispatcher().isRunning() && maxTimeout > 0 {
-		log.Log(log.ShimDispatcher).Info("waiting for dispatcher to be stopped",
-			zap.Int("remainingSeconds", maxTimeout))
-		time.Sleep(1 * time.Second)
-		maxTimeout--
+	stopWait := make(chan struct{})
+	defer close(stopWait)
+
+	go func() {
+		getDispatcher().stopped.Wait()
+		stopWait <- struct{}{}
```

Comment:
We can close the channel directly here.

Reply:
```go
	go func() {
		defer close(stopWait)
		getDispatcher().stopped.Wait()
	}()
```

Reply:
That would mean double closure, which results in a panic. Closing inside `Stop()` seems to be a better place to me.

Reply:
> That would mean double closure

pardon me. why it causes double closure if we remove the `close(stopWait)` from `Stop()`.

```go
	stopWait := make(chan struct{})
        // defer close(stopWait) move this into goroutine

	go func() {
		defer close(stopWait)
		getDispatcher().stopped.Wait()
	}()

	// wait until the main event loop stops properly
	select {
	case <-stopWait:
		break
	case <-time.After(5 * time.Second):
		log.Log(log.ShimDispatcher).Info("dispatcher did not stop in time")
		break
	}
```

Reply:
> pardon me. why it causes double closure if we remove the close(stopWait) from Stop().

In that case it doesn't. I just checked and a general principle is that channel shouldn't be closed on a receiver side. I don't think it's a big deal, but ok, I'll change it.